README.md
# About

This is a fork of original project: 
https://github.com/phpvirtualbox/phpvirtualbox

Branch virtualbox-7 rebased with fork branch: https://github.com/BartekSz95/phpvirtualbox

To get latest fixes with supporting vbox 7.0 release candidate.

### Note ###
Use this fork on your own risk

------------------------------------------------------------------------------
## ORIGINAL README CONTENT ##

# About

phpVirtualBox is from 2017 maintained by Smart Guide Pty Ltd (tudor at smartguide dot com dot au)

with support from various contributors (see https://github.com/phpvirtualbox/phpvirtualbox/graphs/contributors)

Originally Copyright (C) 2015 Ian Moore (imoore76 at yahoo dot com)

FREE, WITHOUT WARRANTY:

All files of this program (phpVirtualBox) are distributed under the
terms contained in the LICENSE.txt file in this folder unless otherwise
specified in an individual source file. By using this software, you are
agreeing to the terms contained therein. If you have not received and read
the license file, or do not agree with its conditions, please cease using
this software immediately and remove any copies you may have in your
possession.

# Installation from Zip file

1) Download zip file from GitHub project site: https://github.com/phpvirtualbox/phpvirtualbox/archive/master.zip

2) Unzip the zipfile into a folder accessible by your web server

3) Rename config.php-example to config.php and edit as needed.

Read the wiki for more information : https://github.com/phpvirtualbox/phpvirtualbox/wiki

# Post installation

Default login is username: admin password: admin

Please report bugs / feature requests to GitHub
https://github.com/phpvirtualbox/phpvirtualbox/issues

# Password Recovery

Rename the file recovery.php-disabled to recovery.php, navigate to it in
your web browser, and follow the instructions presented.


CHANGELOG.txt
-------------------------------------
    7.0-0 RC1
-------------------------------------
Support for VirtualBox 7.0 and PHP8.2 - BartekSz95
Fixed clone option and encryption/decryption - BartekSz95
Fixes for few small bugs - BartekSz95
Fix fileTree listing and extension filtering - derekschrock
Fix api.php exceptions handling bug - miceZipper


-------------------------------------
    6.0-0
-------------------------------------
Upgrade to Virtualbox 6.0


-------------------------------------
    5.2-1   2018-11-09
-------------------------------------
Various snapshot fixes and improvements - @Adam5Wu
Add a configuration variable to set the VRDE address - @rosorio
New icons to match Desktop Application - @der-eismann
Fix Import/Export - @tom077
Russian translation fixes - @SvetlanaTkachenko
Viewport and URL fixes - @krzysztof113
Convert to unix file format to reduce size


-------------------------------------
    5.2-0   2018-02-16
-------------------------------------
Long overdue post 5.0 support (5.2 at this stage)
PHP 7.1 support thanks to @mikedld
New Czech localization thanks to @p-bo
MySQL autho module thanks to @tioigor96
Various bugfixes thanks to @tschoening and @ashway83
Huge support effort from @tom077!
Thanks to everyone who's helped bring this project back up to speed!

-------------------------------------
    5.0-5   2016-01-08
-------------------------------------
PHP 7 support

-------------------------------------
    5.0-4   2015-12-07
-------------------------------------
Fixed bug in file / folder browser when $browserRestrictFolders is
set.
https://sourceforge.net/p/phpvirtualbox/bugs/50/

Fixed bug in host network interface sorting.
https://sourceforge.net/p/phpvirtualbox/bugs/36/

Fixed bug where phpVirtualBox installations on the same server may
share session cookies in certain PHP configurations.

Fixed bug where noPreview=true; was ignored in settings.
https://sourceforge.net/p/phpvirtualbox/bugs/38/

-------------------------------------
    5.0-3   2015-09-09
-------------------------------------
Fixed file browser header content type causing blank file / folder
contents
https://sourceforge.net/p/phpvirtualbox/bugs/49/

Fixed selection highlighting in file / folder browser

Fixed all VMs in a group starting on the same RDP port
https://sourceforge.net/p/phpvirtualbox/bugs/46/

Fixed missing en.xml language file
https://sourceforge.net/p/phpvirtualbox/bugs/45/

-------------------------------------
    5.0-2   2015-08-12
-------------------------------------
Translation fixes

Fixed error when creating host-only networks

Fixed host-only network DHCP server settings

Fix USB device attaching / detaching from Settings menu of a
running VM.

-------------------------------------
    5.0-1   2015-08-05
-------------------------------------
Fix paths in recovery.php-disabled

Fix Content-type header in api.php

Fix add medium button context menu in VM settings

Fix RDP link

Fix bug when changing password

Removed unused OnAdditionsStateChanged event

Change in the way Windows drive list is obtained

Fixed errors manipulating groups when $phpVboxGroups is set

-------------------------------------
    5.0-0	2015-07-18
-------------------------------------
VirtualBox 5.0 compatibility.

Progress operations are now listed in order of creation from top
top bottom.
 
-------------------------------------
	4.3-3	2015-04-04
-------------------------------------
Added 2d/3d video acceleration configuration options in VM
settings

Fixed guest OS icon for Linux 2.6 (32-bit)

Added $eventListenerTimeout configuration option.

Upgraded jQuery to 1.11.2 and jQuery-ui to 1.11.4

-------------------------------------
	4.3-2	2014-12-16
-------------------------------------

Updated documentation to remove all googlecode.com references.

Changing the guest OS type now updates LongMode CPU Feature of VM.
https://sourceforge.net/p/phpvirtualbox/bugs/31/

Removed phptest.php and related check to see if PHP parsing works
correctly. This gives a false negative result in some browsers
/ web server configurations.
https://sourceforge.net/p/phpvirtualbox/bugs/32/

Copied missing image from VirtualBox src.

Fixed bug that would incorrectly determine the name of some images
causing 404s in web server logs and broken images.

Fixed bug in global NAT network settings that would not save the
'Active' checkbox setting.

-------------------------------------
	4.3-1	2014-01-02
-------------------------------------

Fixed bug that would not stop the Global Preferences dialog from
being displayed when $noAuth was enabled or some other auth method
was set in config.php
https://sourceforge.net/p/phpvirtualbox/bugs/7/

Fixed JavaScript error when $enforceVMOwnership is enabled
https://sourceforge.net/p/phpvirtualbox/bugs/9/

Updated the links to point to new Sourceforge home instead of
googlecode
http://sourceforge.net/p/phpvirtualbox/bugs/11/

Updated guest OS type icons

Allow SCSI and SAS controllers to have CD/DVD drives

-------------------------------------
	4.3-0	2013-10-17
-------------------------------------
VirtualBox 4.3.x support.

Fixed behavior when a VM is inaccessible.

Fixed Clone VM Wizard bug that would not keep MAC addresses even
when 'Reinitialize MAC address...' was not set
https://sourceforge.net/p/phpvirtualbox/bugs/3/

Fixed bug that would not allow a VM to be powered off if it stuck
https://sourceforge.net/p/phpvirtualbox/bugs/5/

-------------------------------------
	4.2-8	2013-10-15
-------------------------------------
Updated media menus to indicate which media is already mounted
by the given virtual machine.

Minor UI enhancements.

Fixed bug when collapsing and expanding the Preview box in VM
details sections that would leave an arrow image drawn over the
preview image.

Removed bug workarounds for old version of Opera.

Minor translation bug fixes.

Added VNC viewer for VirtualBox-OSE with VNC support
(experimental).


endpoints/lib/language.php
<?php
/**
 * __vbox_language class and trans() function
 *
 * @author Ian Moore (imoore76 at yahoo dot com)
 * @copyright Copyright (C) 2010-2015 Ian Moore (imoore76 at yahoo dot com)
 * @version $Id: language.php 591 2015-04-11 22:40:47Z imoore76 $
 * @see languages/languages.txt
 * @package phpVirtualBox
 *
 */

global $_vbox_language;

// Settings contains language
require_once(dirname(__FILE__).'/config.php');
require_once(dirname(__FILE__).'/utils.php');

define('VBOX_BASE_LANG_DIR', dirname(dirname(dirname(__FILE__))) .'/languages');

/**
 * Language class. Parses language file and stores translations in an array.
 *
 * @author Ian Moore (imoore76 at yahoo dot com)
 * @copyright Copyright (C) 2010-2015 Ian Moore (imoore76 at yahoo dot com)
 * @version $Id: language.php 591 2015-04-11 22:40:47Z imoore76 $
 * @package phpVirtualBox
 *
 * @global __vbox_language $GLOBALS['_vbox_language'] global __vbox_language instance
 * @name $_vbox_language
*/
class __vbox_language {

	/**
	 * Static language data used for translations
	 * @var array
	 */
	static $langdata = null;

	/**
	 *
	 * Constructor parses language file and stores translations.
	 */
	function __construct() {

		# Already initialized?
		if(is_array(self::$langdata)) return;

		self::$langdata = array();

		$settings = new phpVBoxConfigClass();
		$lang = strtolower($settings->language);

		if(@$_COOKIE['vboxLanguage']) {
			$lang = str_replace(array('/','\\','.'),'',$_COOKIE['vboxLanguage']);
		}

		// File as specified
		if($lang && file_exists(VBOX_BASE_LANG_DIR.'/source/'.$lang.'.dat')) {
			@define('VBOXLANG', $lang);

		// No lang file found
		} else {
			$lang = 'en';
			@define('VBOXLANG', $lang);
			self::$langdata['contexts'] = array();
			return;
		}


		self::$langdata = unserialize(@file_get_contents(VBOX_BASE_LANG_DIR.'/source/'.$lang.'.dat'));

		$xmlObj = simplexml_load_string(@file_get_contents(VBOX_BASE_LANG_DIR.'/'.$lang.'.xml'));
		$arrXml = $this->objectsIntoArray($xmlObj);

        if(!array_key_exists('context',$arrXml)) return;

		$lang = array();
		if(!@$arrXml['context'][0]) $arrXml['context'] = array($arrXml['context']);
		foreach($arrXml['context'] as $c) {

			if(!is_array($c) || !@$c['name']) continue;
			if(!@$c['message'][0]) $c['message'] = array($c['message']);

		   	$lang['contexts'][$c['name']] = array();
		   	$lang['contexts'][$c['name']]['messages'] = array();

			foreach($c['message'] as $m) {

		       if(!is_array($m)) continue;

		       $s = $m['source'];
		       unset($m['source']);
		       $lang['contexts'][$c['name']]['messages'][$s] = $m;
	    	}
		}
		self::$langdata = array_merge_recursive(self::$langdata, $lang);
	}

	/**
	 * Translate text.
	 * @param string $item message to translate
	 * @param string $context context in which the translation should be performed
	 */
	function trans($item,$context='phpVirtualBox') {
		$t = @self::$langdata['contexts'][$context]['messages'][$item]['translation'];
		return ($t ? $t : $item);
	}

	/**
	 *
	 * Converts objects into array. Called from class constructor.
	 * @param array|object $arrObjData object or array to convert to array
	 * @param array $arrSkipIndices array of indices to skip
	 * @return array
	 */
	function objectsIntoArray($arrObjData, $arrSkipIndices = array())
	{
	    $arrData = array();

	    // if input is object, convert into array
	    if (is_object($arrObjData)) {
	        $arrObjData = get_object_vars($arrObjData);
	    }

	    if (is_array($arrObjData)) {
	        foreach ($arrObjData as $index => $value) {
	            if (is_object($value) || is_array($value)) {
	                $value = $this->objectsIntoArray($value, $arrSkipIndices); // recursive call
	            }
	            if (in_array($index, $arrSkipIndices)) {
	                continue;
	            }
	            $arrData[$index] = $value;
	        }
	    }
	    return $arrData;
	}

}

/**
 * Translate text. If $GLOBALS['_vbox_language'] is not set, create a new
 * instance and pass params to its trans() method
 * @param string $msg message to translate
 * @param string $context context in which the translation should be performed
 * @uses $_vbox_language
 * @return string
 */
function trans($msg,$context='phpVirtualBox') {
	if(!is_object($GLOBALS['_vbox_language'])) $GLOBALS['_vbox_language'] = new __vbox_language();
	return $GLOBALS['_vbox_language']->trans($msg,$context);
}


endpoints/api.php
<?php
/**
 * Main API interface between JavaScript ajax calls and PHP functions.
 * Accepts JSON, POST data or simple GET requests, and returns JSON data.
 *
 * @author Ian Moore (imoore76 at yahoo dot com)
 * @copyright Copyright (C) 2010-2015 Ian Moore (imoore76 at yahoo dot com)
 * @version $Id: api.php 596 2015-04-19 11:50:53Z imoore76 $
 * @package phpVirtualBox
 * @see vboxconnector
 * @see vboxAjaxRequest
 *
 * @global array $GLOBALS['response'] resopnse data sent back via json
 * @name $response
*/

# Turn off PHP errors
error_reporting(E_ALL & ~E_NOTICE & ~E_STRICT & ~E_WARNING);


//Set no caching
header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
header("Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0");
header("Pragma: no-cache");

require_once(dirname(__FILE__).'/lib/config.php');
require_once(dirname(__FILE__).'/lib/utils.php');
require_once(dirname(__FILE__).'/lib/vboxconnector.php');

// Init session
global $_SESSION;

/*
 * Clean request
 */
$request = clean_request();


global $response;
$response = array('data'=>array('responseData'=>array()),'errors'=>array(),'persist'=>array(),'messages'=>array());

/*
 * Built-in requests
 */
$vbox = null; // May be set during request handling

/**
 * Main try / catch. Logic dictated by incoming 'fn' request
 * parameter.
 */
try {

	/* Check for password recovery file */
	if(file_exists(dirname(dirname(__FILE__)).'/recovery.php')) {
		throw new Exception('recovery.php exists in phpVirtualBox\'s folder. This is a security hazard. phpVirtualBox will not run until recovery.php has been renamed to a file name that does not end in .php such as <b>recovery.php-disabled</b>.',vboxconnector::PHPVB_ERRNO_FATAL);
	}

	/* Check for PHP version */
	if (!version_compare(PHP_VERSION, '5.2.0', '>=')) {
		throw new Exception('phpVirtualBox requires PHP >= 5.2.0, but this server is running version '. PHP_VERSION .'. Please upgrade PHP.');
	}

	# Only valid function chars
	$request['fn'] = preg_replace('[^a-zA-Z0-9_-]', '', $request['fn']);

	/* Check for function called */
	switch($request['fn']) {

		/*
		 * No method called
		 */
	    case '':
	    	throw new Exception('No method called.');
	    	break;

		/*
		 * Return phpVirtualBox's configuration data
		 */
		case 'getConfig':

			$settings = new phpVBoxConfigClass();
			$response['data']['responseData'] = get_object_vars($settings);
			$response['data']['responseData']['host'] = parse_url($response['data']['responseData']['location']);
			$response['data']['responseData']['host'] = $response['data']['responseData']['host']['host'];
			$response['data']['responseData']['phpvboxver'] = @constant('PHPVBOX_VER');

			// Session
			session_init();

			// Hide credentials
			unset($response['data']['responseData']['username']);
			unset($response['data']['responseData']['password']);
			foreach($response['data']['responseData']['servers'] as $k => $v)
				$response['data']['responseData']['servers'][$k] = array('name'=>$v['name']);

			// Vbox version
			$vbox = new vboxconnector();
			$response['data']['responseData']['version'] = $vbox->getVersion();
			$response['data']['responseData']['hostOS'] = $vbox->vbox->host->operatingSystem;
			$response['data']['responseData']['DSEP'] = $vbox->getDsep();
			$response['data']['responseData']['groupDefinitionKey'] = ($settings->phpVboxGroups ? vboxconnector::phpVboxGroupKey : 'GUI/GroupDefinitions');

			$response['data']['success'] = true;

			break;

		/*
		 *
		 * USER FUNCTIONS FOLLOW
		 *
		 */

		/*
		 * Pass login to authentication module.
		 */
		case 'login':


			// NOTE: Do not break. Fall through to 'getSession
			if(!$request['params']['u'] || !$request['params']['p']) {
				break;
			}

			// Session
			session_init(true);

			$settings = new phpVBoxConfigClass();

			// Try / catch here to hide login credentials
			try {
				$settings->auth->login($request['params']['u'], $request['params']['p']);
			} catch(Exception $e) {
				throw new Exception($e->getMessage(), $e->getCode());
			}

			// We're done writing to session
			if(function_exists('session_write_close'))
				@session_write_close();



		/*
		 * Return $_SESSION data
		 */
		case 'getSession':

			$settings = new phpVBoxConfigClass();
			if(method_exists($settings->auth,'autoLoginHook'))
			{
				// Session
				session_init(true);

				$settings->auth->autoLoginHook();

				// We're done writing to session
				if(function_exists('session_write_close'))
					@session_write_close();

			} else {

				session_init();

			}


			$response['data']['responseData'] = $_SESSION;
			$response['data']['success'] = true;
			break;

		/*
		 * Change phpVirtualBox password. Passed to auth module's
		 * changePassword method.
		 */
		case 'changePassword':

			// Session
			session_init(true);

			$settings = new phpVBoxConfigClass();
			$response['data']['success'] = $settings->auth->changePassword($request['params']['old'],
					                                                       $request['params']['new']);

			// We're done writing to session
			if(function_exists('session_write_close'))
				@session_write_close();

			break;

		/*
		 * Get a list of phpVirtualBox users. Passed to auth module's
		 * getUsers method.
		 */
		case 'getUsers':

			// Session
			session_init();

			// Must be an admin
			if(!$_SESSION['admin']) break;

			$settings = new phpVBoxConfigClass();
			$response['data']['responseData'] = $settings->auth->listUsers();
			$response['date']['success'] = true;

			break;

		/*
		 * Remove a phpVirtualBox user. Passed to auth module's
		 * deleteUser method.
		 */
		case 'delUser':

			// Session
			session_init();

			// Must be an admin
			if(!$_SESSION['admin']) break;

			$settings = new phpVBoxConfigClass();
			$settings->auth->deleteUser($request['params']['u']);

			$response['data']['success'] = true;
			break;

		/*
		 * Edit a phpVirtualBox user. Passed to auth module's
		 * updateUser method.
		 */
		case 'editUser':

			$skipExistCheck = true;
			// Fall to addUser

		/*
		 * Add a user to phpVirtualBox. Passed to auth module's
		 * updateUser method.
		 */
		case 'addUser':

			// Session
			session_init();

			// Must be an admin
			if(!$_SESSION['admin']) break;

			$settings = new phpVBoxConfigClass();
			$settings->auth->updateUser($request['params'], @$skipExistCheck);

			$response['data']['success'] = true;
			break;

		/*
		 * Log out of phpVirtualBox. Passed to auth module's
		 * logout method.
		 */
		case 'logout':

			// Session
			session_init(true);

			$vbox = new vboxconnector();
			$vbox->skipSessionCheck = true;

			$settings = new phpVBoxConfigClass();
			$settings->auth->logout($response);

			session_destroy();

			$response['data']['success'] = true;

			break;


		/*
		 * If the above cases did not match, assume it is a request
		 * that should be passed to vboxconnector.
		 */
		default:

			$vbox = new vboxconnector();


			/*
			 * Every 1 minute we'll check that the account has not
			 * been deleted since login, and update admin credentials.
			 */
			if($_SESSION['user'] && ((intval($_SESSION['authCheckHeartbeat'])+60) < time())) {

				// init session and keep it open
				session_init(true);
				$vbox->settings->auth->heartbeat($vbox);

				// We're done writing to session
				if(function_exists('session_write_close'))
					@session_write_close();

			} else {

				// init session but close it
				session_init();

			}

			/*
			 *  Persistent request data
			 */
			if(is_array($request['persist'])) {
				$vbox->persistentRequest = $request['persist'];
			}


			/*
			 * Call to vboxconnector
			 */
			$vbox->{$request['fn']}($request['params'],array(&$response));


			/*
			 * Send back persistent request in response
			*/
			if(is_array($vbox->persistentRequest) && count($vbox->persistentRequest)) {
				$response['data']['persist'] = $vbox->persistentRequest;
			}
			break;

	} // </switch()>

/*
 * Catch all exceptions and populate errors in the
 * JSON response data.
 */
} catch (Exception $e) {

	// Just append to $vbox->errors and let it get
	// taken care of below
    if(!isset($vbox)){
        $vbox = new stdClass();
    }
	if(!$vbox || !$vbox->errors) {
		$vbox->errors = array();
	}
	$vbox->errors[] = $e;
}


// Add any messages
if($vbox && isset($vbox->messages)?count($vbox->messages):false) {
	foreach($vbox->messages as $m)
		$response['messages'][] = 'vboxconnector('.$request['fn'] .'): ' . $m;
}
// Add other error info
if($vbox && $vbox->errors) {

	foreach($vbox->errors as $e) { /* @var $e Exception */

		ob_start();
		print_r($e);
		$d = ob_get_contents();
		ob_end_clean();

		# Add connection details to connection errors
		if($e->getCode() == vboxconnector::PHPVB_ERRNO_CONNECT && isset($vbox->settings))
			$d .= "\n\nLocation:" . $vbox->settings->location;

		$response['messages'][] = htmlentities($e->getMessage()). htmlentities(' '. $details);

		$response['errors'][] = array(
			'error'=> ($e->getCode() & vboxconnector::PHPVB_ERRNO_HTML ? $e->getMessage() : htmlentities($e->getMessage())),
			'details'=>htmlentities($d),
			'errno'=>$e->getCode(),
			// Fatal errors halt all processing
			'fatal'=>($e->getCode() & vboxconnector::PHPVB_ERRNO_FATAL),
			// Connection errors display alternate servers options
			'connection'=>($e->getCode() & vboxconnector::PHPVB_ERRNO_CONNECT)
		);
	}
}

/*
 * Return response as JSON encoded data or use PHP's
 * print_r to dump data to browser.
 */
if(isset($request['printr'])) {
	print_r($response);
} else {
    header('Content-type: application/json');
	echo(json_encode($response));
}

